src/main/scala/nl/knaw/dans/easy/multideposit/MultiDepositParser.scala

```diff
@@ -17,79 +17,77 @@ package nl.knaw.dans.easy.multideposit
+object MultiDepositParser extends DebugEnhancedLogging {
 
   type CsvValues = List[(String, String)]
 
-  val log = LoggerFactory.getLogger(MultiDepositParser.getClass)
-
   /**
-    * Transforms a `File` into a stream of `Datasets` by parsing the csv.
-    *
-    * @param file the csv file to be parsed
-    * @return a stream of datasets coming from the file
-    */
-  def parse(file: File): Observable[Datasets] = {
-    Observable(subscriber => {
-      log.debug("Start parsing SIP Instructions at {}", file)
-      Try {
+   * Transforms a `File` into a `Datasets` instance by parsing the csv.
+   *
+   * @param file the csv file to be parsed
+   * @return a collection of datasets coming from the file
+   */
+  def parse(file: File): Try[Datasets] = {
+    debug(s"Start parsing SIP Instructions at $file")
+
+    for {
+      content <- Try {
         val rawContent = Source.fromFile(file)(Codec(encoding)).mkString
         val parser = CSVParser.parse(rawContent, CSVFormat.RFC4180)
-        val output = parser.getRecords.filter(r => r.size() > 0 && !r.get(0).isBlank).map(_.toList)
-        validateDatasetHeaders(output.head)
-          .map(_ => {
-            val csvData = output.tail.map(output.head zip _)
-            log.debug("Successfully loaded CSV file")
-            Observable.from(csvData)
-              .zipWithIndex
-              .map { case (xs, i) => (xs, i + 2) } // +2 because we want the second line (element 0) to have index 2
-              .foldLeft(new Datasets)((dss, tuple) => updateDatasets(dss, tuple))
-          })
-          .onError(Observable.error)
-          .subscribe(subscriber)
-      }.onError(subscriber.onError)
-    })
+        for {
+          record <- parser.getRecords.asScala
+          if record.size() > 0
+          if !record.get(0).isBlank
+        } yield record.asScala.toList
+      }
+      _ <- validateDatasetHeaders(content.head)
+    } yield {
+      debug("Successfully loaded CSV file")
+      content.tail
+        .map(content.head.zip(_))
+        .zipWithIndex
+        .foldRight(Datasets) {
+          // +2 because we want the second line (element 0) to have index 2
+          case ((values: CsvValues, row: Int), dss: Datasets) => updateDatasets(dss, values, row + 1)
+        }
+    }
   }
 
   /**
-    * Tests whether the given list of headers is valid. If so, `Success(Unit)` is returned,
-    * else a `Failure` with a detailed `ActionException` is returned.
-    *
-    * @param headers the headers to be validated
-    * @return `Success` if the `headers` are valid, `Failure` if the `headers` are invalid
-    */
+   * Tests whether the given list of headers is valid. If so, `Success(Unit)` is returned,
+   * else a `Failure` with a detailed `ActionException` is returned.
+   *
+   * @param headers the headers to be validated
+   * @return `Success` if the `headers` are valid, `Failure` if the `headers` are invalid
+   */
+  // TODO valid headers listing as sub-command???
```

Comment:
neither, IMO. The one you refer to maps to an object; this project maps its values to another kind of datastructure (`HashMap`)